lib/gemstash/S3.rb

```diff
@@ -0,0 +1,247 @@
+require "gemstash"
+require "digest"
+require "fileutils"
+require "yaml"
+require "aws-sdk-s3"
+require "pathname"
+
+module Gemstash
+  class S3
+    extend Gemstash::Env::Helper
+    attr_reader :folder, :client, :bucket
+
+    VERSION = 1
+    class VersionTooNew < StandardError
+      def initialize(folder, version)
+        super("Gemstash storage version #{Gemstash::S3::VERSION} does " \
```

Comment:
Can you explain what this error means?

Reply:
If you take a look at the initialization at line 23, it is used to make sure that when the S3 Storage class is originally initialized, the resource version is checked. If root = false, or if the initialization isn't the original initialization, but merely adding on to a previous initialization, then it is not checked. I'm unsure how the addition of a root variable helps, but I think it improves performance so that you don't have to check the version if it isn't the root initialization.